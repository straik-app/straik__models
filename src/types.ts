
export type ReadOnlyFields<T> = {
  [Key in keyof T]: true | undefined;
};

export const True = true as const;
