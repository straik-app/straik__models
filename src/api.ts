export type Method = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'OPTIONS' | 'HEAD' | 'DELETE';

export interface RouteOptions {
  version?: string;
  idColumn?: string;
  suffix?: string;
  method?: Method;
}

export interface Route {
  resource: string;
  version: string;
  method: Method;
  path: string;
  label: string;
}

/**
 * Builds a Route object
 *
 * @param resource
 * @param options
 */
export const RouteBuilder = (resource: string, options: RouteOptions = {}): Route => {
  const version = options.version || 'v1';

  const components = [
    '',
    version,
    resource,
  ];

  if (options.idColumn) {
    components.push(`:${options.idColumn}`);
  }

  if (options.suffix) {
    components.push(options.suffix);
  }

  return {
    resource,
    version,
    method: options.method || 'GET',
    path: components.join('/'),

    /**
     * Generates a label for the route
     */
    get label() {
      return `${this.method} ${this.path}`;
    },
  };
};

/**
 * Builds a Retrieve Route object
 *
 * @param resource
 * @param options
 */
export const RetrieveRoute = (resource: string, options: RouteOptions = {}): Route => {
  return RouteBuilder(resource, { idColumn: 'uuid', ...options });
};

/**
 * Builds a List Route object
 *
 * @param resource
 * @param options
 */
export const ListRoute = (resource: string, options: RouteOptions = {}): Route => {
  return RouteBuilder(resource, options);
};

/**
 * Builds a Create Route object
 *
 * @param resource
 * @param options
 */
export const CreateRoute = (resource: string, options: RouteOptions = {}): Route => {
  return RouteBuilder(resource, { method: 'POST', ...options });
};

/**
 * Builds an Update Route object
 *
 * @param resource
 * @param options
 */
export const UpdateRoute = (resource: string, options: RouteOptions = {}): Route => {
  return RouteBuilder(resource, { method: 'PATCH', idColumn: 'uuid', ...options });
};

/**
 * Builds a Destroy Route object
 *
 * @param resource
 * @param options
 */
export const DestroyRoute = (resource: string, options: RouteOptions = {}): Route => {
  return RouteBuilder(resource, { method: 'DELETE', idColumn: 'uuid', ...options });
};

/**
 * Straik REST API Routes
 */
export const Routes = {
  home: {
    resource: '',
    version: 'v1',
    method: 'GET',
    path: '/',
    label: 'GET /',
  },
  activities: {
    retrieve: RetrieveRoute('activities'),
    list: ListRoute('activities'),
    create: CreateRoute('activities'),
    update: UpdateRoute('activities'),
    destroy: DestroyRoute('activities'),
  },
  spaces: {
    retrieve: RetrieveRoute('spaces'),
    list: ListRoute('spaces'),
    create: CreateRoute('spaces'),
    update: UpdateRoute('spaces'),
    destroy: DestroyRoute('spaces'),
  },
  status: {
    retrieve: RouteBuilder('status'),
  },
  tags: {
    retrieve: RetrieveRoute('tags'),
    list: ListRoute('tags'),
    create: CreateRoute('tags'),
    update: UpdateRoute('tags'),
    destroy: DestroyRoute('tags'),
  },
  users: {
    retrieve: RetrieveRoute('users'),
    list: ListRoute('users'),
    auth: RouteBuilder('users', { suffix: 'auth', method: 'POST' }),
    update: UpdateRoute('users'),
    destroy: DestroyRoute('users'),
  },
};
