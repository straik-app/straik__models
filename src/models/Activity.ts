import { v4 as uuid4 } from 'uuid';
import * as zod from 'zod';

import { TimestampField } from '../fields/Timestamp';
import { True } from '../types';

export const ActivityResponseOmit = {};
export const ActivityCreateOmit = { created: True, user_uuid: True };
export const ActivityUpdateOmit = { ...ActivityCreateOmit, uuid: True };


export const ActivityModel = zod.object({
  /** Activity UUID */
  uuid: zod.string().uuid().default(() => uuid4()),

  /** Related user UUID */
  user_uuid: zod.string().uuid().nonempty(),

  /** Related space UUID */
  space_uuid: zod.string().uuid().nonempty(),

  /** Title */
  title: zod.string().nonempty(),

  /** Activity description. Accepts markdown */
  description: zod.string(),

  /** When the activity begins */
  begins: TimestampField,

  /** When the activity ends */
  ends: TimestampField,

  /** Creation timestamp */
  created: TimestampField,
});

export const ActivityResponseModel = ActivityModel.omit(ActivityResponseOmit);

export const ActivityCreateModel = ActivityModel.omit(ActivityCreateOmit);

export const ActivitUpdateModel = ActivityCreateModel.omit(ActivityUpdateOmit).partial();

export type Activity = zod.infer<typeof ActivityModel>;
export type ActivityCreate = zod.infer<typeof ActivityCreateModel>;
export type ActivityUpdate = zod.infer<typeof ActivitUpdateModel>;
export type ActivityResponse = zod.infer<typeof ActivityResponseModel>;
