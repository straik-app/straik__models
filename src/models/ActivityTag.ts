import { v4 as uuid4 } from 'uuid';
import * as zod from 'zod';

import { TimestampField } from '../fields/Timestamp';

export const ActivityTagModel = zod.object({
  /** Activity UUID */
  uuid: zod.string().uuid().default(() => uuid4()),

  /** Related activity UUID */
  activity_uuid: zod.string().uuid().nonempty(),

  /** Related tag UUID */
  tag_uuid: zod.string().uuid().nonempty(),

  /** Activity description. Accepts markdown */
  description: zod.string(),

  /** Creation timestamp */
  created: TimestampField,
});

export type ActivityTag = zod.infer<typeof ActivityTagModel>;
