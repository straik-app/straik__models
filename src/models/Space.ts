import { v4 as uuid4 } from 'uuid';
import * as zod from 'zod';

import { SlugField } from '../fields/Slug';
import { TimestampField } from '../fields/Timestamp';
import { True } from '../types';

export const SpaceResponseOmit = {};
export const SpaceCreateOmit = { created: True, slug: True, user_uuid: True };
export const SpaceUpdateOmit = { ...SpaceCreateOmit, uuid: True };

export const SpaceModel = zod.object({
  /** Space UUID */
  uuid: zod.string().uuid().default(() => uuid4()),

  /** Related user UUID */
  user_uuid: zod.string().uuid().nonempty(),

  /** Title */
  title: zod.string().nonempty(),

  /** Slug */
  slug: SlugField.nonempty(),

  /** Space description */
  description: zod.string(),

  /** Creation timestamp */
  created: TimestampField.optional(),
});

export const SpaceCreateModel = SpaceModel.omit(SpaceCreateOmit);

export const SpaceUpdateModel = SpaceModel.omit(SpaceUpdateOmit).partial();

export const SpaceResponseModel = SpaceModel.omit(SpaceResponseOmit);

export type Space = zod.infer<typeof SpaceModel>;
export type SpaceCreate = zod.infer<typeof SpaceCreateModel>;
export type SpaceUpdate = zod.infer<typeof SpaceUpdateModel>;
export type SpaceResponse = zod.infer<typeof SpaceResponseModel>;
