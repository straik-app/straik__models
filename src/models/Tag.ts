import { v4 as uuid4 } from 'uuid';
import * as zod from 'zod';

import { SlugField } from '../fields/Slug';
import { TimestampField } from '../fields/Timestamp';
import { True } from '../types';

export const TagResponseOmitFields = {};
export const TagCreateOmitFields = { user_uuid: True, slug: True, created: True };
export const TagUpdateOmitFields = { ...TagCreateOmitFields, uuid: True };

export const TagModel = zod.object({
  /** Tag UUID */
  uuid: zod.string().uuid().default(() => uuid4()),

  /** Related user UUID */
  user_uuid: zod.string().uuid().nonempty(),

  /** Title */
  title: zod.string().nonempty(),

  /** Slug */
  slug: SlugField.nonempty(),

  /** Tag description */
  description: zod.string(),

  /** Creation timestamp */
  created: TimestampField,
});

export const TagCreateModel = TagModel.omit(TagCreateOmitFields);
export const TagUpdateModel = TagModel.omit(TagUpdateOmitFields).partial();
export const TagResponseModel = TagModel.omit(TagResponseOmitFields);

export type Tag = zod.infer<typeof TagModel>;
export type TagCreate = zod.infer<typeof TagCreateModel>;
export type TagUpdate = zod.infer<typeof TagUpdateModel>;
export type TagResponse = zod.infer<typeof TagResponseModel>;
