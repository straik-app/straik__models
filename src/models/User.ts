import { v4 as uuid4 } from 'uuid';
import * as zod from 'zod';

import { TimestampField } from '../fields/Timestamp';
import { True } from '../types';

/**
 * User subscriptions
 */
export enum UserSubscriptions {
  FREE = 'FREE',
  PRO = 'PRO',
}

export const UserCreateOmitFields = {
  admin: True,
  subscription: True,
  verified: True,
  created: True,
};
export const UserUpdateOmitFields = { ...UserCreateOmitFields, uuid: True };
export const UserResponseOmitFields = { password: True, verified: True };

/**
 * User DB model
 */
export const UserModel = zod.object({
  /** User UUID */
  uuid: zod.string().uuid().default(() => uuid4()),

  /** User full name */
  full_name: zod.string().nonempty(),

  /** User preferred or short name */
  pref_name: zod.string().nonempty(),

  /** User email */
  email: zod.string().email(),

  /**
   * User password
   *
   * This value will be a salted hash in the backend
   * and missing in the frontend.
   */
  password: zod.string().optional(),

  /** Whether the user is an admin */
  admin: zod.boolean().default(() => false),

  /** Type of subscription */
  subscription: zod.nativeEnum(UserSubscriptions).default(() => UserSubscriptions.FREE),

  /** Whether the user verified their account */
  verified: zod.boolean().default(() => false),

  /** Creation timestamp */
  created: TimestampField,
});

export const UserCreateModel = UserModel.omit(UserCreateOmitFields).strip();
export const UserUpdateModel = UserModel.omit(UserUpdateOmitFields).strip().partial();

/**
 * User API response
 */
export const UserResponseModel = UserModel.omit(UserResponseOmitFields).strip();

export type User = zod.infer<typeof UserModel>;
export type UserCreate = zod.infer<typeof UserCreateModel>;
export type UserUpdate = zod.infer<typeof UserUpdateModel>;
export type UserResponse = zod.infer<typeof UserResponseModel>;
