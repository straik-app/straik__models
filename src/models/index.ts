
export * from './Activity';
export * from './ActivityTag';
export * from './Space';
export * from './Tag';
export * from './User';

export interface Validator<T> {
  parse: (values: unknown) => T
}
