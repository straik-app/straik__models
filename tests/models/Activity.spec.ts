import { ActivityFixtures, ActivityGenerator } from '../../src/fixtures';
import { ActivityModel } from '../../src/models';
import { ModelTest } from '../mixins';

describe('models.Activity', () => {
  ModelTest(ActivityModel, ActivityFixtures, ActivityGenerator);
});
