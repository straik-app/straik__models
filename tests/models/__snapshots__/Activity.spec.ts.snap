// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models.Activity invalidates DEFAULT 1`] = `
"[
  {
    \\"validation\\": \\"uuid\\",
    \\"code\\": \\"invalid_string\\",
    \\"path\\": [
      \\"uuid\\"
    ],
    \\"message\\": \\"Invalid uuid\\"
  },
  {
    \\"validation\\": \\"uuid\\",
    \\"code\\": \\"invalid_string\\",
    \\"path\\": [
      \\"user_uuid\\"
    ],
    \\"message\\": \\"Invalid uuid\\"
  },
  {
    \\"validation\\": \\"uuid\\",
    \\"code\\": \\"invalid_string\\",
    \\"path\\": [
      \\"space_uuid\\"
    ],
    \\"message\\": \\"Invalid uuid\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"number\\",
    \\"path\\": [
      \\"title\\"
    ],
    \\"message\\": \\"Expected string, received number\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"number\\",
    \\"path\\": [
      \\"description\\"
    ],
    \\"message\\": \\"Expected string, received number\\"
  },
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Invalid timestamp\\",
    \\"path\\": [
      \\"begins\\"
    ]
  },
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Invalid timestamp\\",
    \\"path\\": [
      \\"ends\\"
    ]
  },
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Invalid timestamp\\",
    \\"path\\": [
      \\"created\\"
    ]
  }
]"
`;

exports[`models.Activity validates DEFAULT 1`] = `
Object {
  "begins": "2021-03-05T21:14:59.057Z",
  "created": "2021-03-05T21:14:59.057Z",
  "description": "Tutti i miei piani & le mie azioni contro Firenze",
  "ends": "2021-03-05T21:14:59.057Z",
  "space_uuid": "e141f390-b41e-43b3-82d0-280ea664a0a8",
  "title": "Vendetta contro Firenze",
  "user_uuid": "e141f390-b41e-43b3-82d0-280ea664a0a8",
  "uuid": "e141f390-b41e-43b3-82d0-280ea664a0a8",
}
`;
